stdlib/public/core/Random.swift

```diff
@@ -0,0 +1,167 @@
+///         case sunday, monday, tuesday, wednesday, thursday, friday, saturday
+///
+///         static func randomWeekday<G: RandomNumberGenerator>(using generator: inout G) -> Weekday {
+///             return Weekday.allCases.randomElement(using: &generator)!
+///         }
+///
+///         static func randomWeekday() -> Weekday {
+///             return Weekday.randomWeekday(using: &Random.default)
+///         }
+///     }
+///
+/// Conforming to the RandomNumberGenerator Protocol
+/// ================================================
+///
+/// A custom `RandomNumberGenerator` type can have different characteristics
+/// than the default `Random` type. For example, a seedable generator can be
+/// used to generate the same sequence of random values for testing purposes.
+///
+/// To make a custom type conform to the `RandomNumberGenerator` protocol,
+/// implement the required `next()` method. Each call to `next()` must produce
+/// a uniform and independent random value.
+///
+/// Types that conform to `RandomNumberGenerator` should specifically document
+/// the thread safety and quality of the generator.
+public protocol RandomNumberGenerator {
+  /// Returns a value from a uniform, independent distribution of binary data.
+  ///
+  /// - Returns: An unsigned 64-bit random value.
+  mutating func next() -> UInt64
+}
+
+extension RandomNumberGenerator {
+  /// Returns a value from a uniform, independent distribution of binary data.
+  ///
+  /// - Returns: A random value of `T`. Bits are randomly distributed so that
+  ///   every value of `T` is equally likely to be returned.
+  @inlinable
+  public mutating func next<T: FixedWidthInteger & UnsignedInteger>() -> T {
+    if T.bitWidth <= UInt64.bitWidth {
+      return T(truncatingIfNeeded: self.next())
+    }
+
+    let (quotient, remainder) = T.bitWidth.quotientAndRemainder(
+      dividingBy: UInt64.bitWidth
+    )
+    var tmp: T = 0
+
+    for i in 0 ..< quotient {
+      tmp += T(truncatingIfNeeded: self.next()) &<< (UInt64.bitWidth * i)
+    }
+
+    if remainder != 0 {
+      let random = self.next()
+      let mask = UInt64.max &>> (UInt64.bitWidth - remainder)
+      tmp += T(truncatingIfNeeded: random & mask) &<< (UInt64.bitWidth * quotient)
+    }
+
+    return tmp
+  }
+
+  /// Returns a random value that is less than the given upper bound.
+  ///
+  /// - Parameter upperBound: The upper bound for the randomly generated value.
+  /// - Returns: A random value of `T` in the range `0..<upperBound`. Every
+  ///   value in the range `0..<upperBound` is equally likely to be returned.
+  @inlinable
+  public mutating func next<T: FixedWidthInteger & UnsignedInteger>(
+    upperBound: T
+  ) -> T {
+    guard upperBound != 0 else { return 0 }
+    let tmp = (T.max % upperBound) + 1
+    let range = tmp == upperBound ? 0 : tmp
+    var random: T = 0
+
+    repeat {
+      random = self.next()
+    } while random < range
+
+    return random % upperBound
+  }
```

Comment:
Could you add a comment explaining the implementation strategy and the reason you chose it?